jwst/jump/tests/test_miri_first_lastframe.py

```diff
@@ -0,0 +1,82 @@
+"""
+Unit test for jump step to test not using MIRI first frame or last frame if set to DO_NOT_USE
+"""
+import pytest
+import numpy as np
+from jwst.jump import JumpStep
+from jwst import datamodels
+from jwst.datamodels import dqflags, MIRIRampModel
+from jwst.pipeline.collect_pipeline_cfgs import collect_pipeline_cfgs
+
+
+@pytest.fixture(scope='function')
+def science_image():
+    """Generate science image"""
+    # size of integration
+    nints = 1
+    ngroups = 7
+    xsize = 1032
+    ysize = 1024
+
+    # create the data and groupdq arrays
+    # create a JWST datamodel for MIRI data
+    im = MIRIRampModel((nints, ngroups, ysize, xsize))
+    im.meta.instrument.detector = 'MIRIMAGE'
+    im.meta.instrument.name = 'MIRI'
+    im.meta.subarray.name = 'FULL'
+    im.meta.subarray.xstart = 1
+    im.meta.subarray.ystart = 1
+    im.meta.subarray.xsize = 1032
+    im.meta.subarray.ysize = 1024
+
+    im.meta.observation.date = '2019-02-27'
+    im.meta.observation.time = '13:37:18.548'
+    im.meta.date = '2019-02-27T13:37:18.548'
+    im.meta.exposure.readpatt = 'FAST'
+    im.meta.exposure.nframes = ngroups
+    im.meta.exposure.group_time = 2.77504
+
+    # set up the data so that if the first and last group are used in jump
+    # detection it would cause a jump to be detected between group 1-2
+    # and group 6-7
+    # set group 1 to be 10,000
+    im.data[0, 0,:,:] = 10000.0
+    # set groups 1,2 - to be around 30,000
+    im.data[0, 1,:,:] = 30000.0
+    im.data[0, 2,:,:] = 30020.0
+    # set up a jump
+    im.data[0, 3,:,:] = 40040.0
+    im.data[0, 4,:,:] = 40060.0
+    im.data[0, 5,:,:] = 40080.0
+
+    # set group 6 to be 50,000
+    im.data[0, 6,:,:] = 50000.0
+
+    # set the first and last group to DO_NOTUSE
+    im.groupdq[0, 0,:,:] = dqflags.group['DO_NOT_USE']
+    im.groupdq[0, -1,:,:] = dqflags.group['DO_NOT_USE']
+
+    return im
+
+
+def test_miri_first_lastframe(_jail, science_image):
+
+    im = datamodels.open(science_image)
+    xsize = im.meta.subarray.xsize
+    ysize = im.meta.subarray.ysize
+
+    collect_pipeline_cfgs('./config')
+    result = JumpStep.call(im, config_file='config/jump.cfg')
+
+    first_group = result.groupdq[0,0,:,:]
+    last_group = result.groupdq[0,6,:,:]
+    jump_group = result.groupdq[0,3,:,:]
+
+    do_not_use_array = np.zeros((ysize,xsize))
+    jump_array = np.zeros((ysize,xsize))
+    do_not_use_array[:,:] = dqflags.group['DO_NOT_USE']
+    jump_array[:,:] = dqflags.group['JUMP_DET']
+
+    assert np.array_equal(first_group, do_not_use_array)
+    assert np.array_equal(last_group, do_not_use_array)
+    assert np.array_equal(jump_group, jump_array)
```

Comment:
Since this PR only changes the `twopoint_difference.find_crs()` function, I would suggest putting the new test with the other 29 tests for this function already in the module `test_twopoint_difference.py`.  And instead of testing the whole step (with all the overhead of a step), just test the function, like the other tests.  And use the same fixture already defined.  I.e. something like:
```

def test_first_last_frame(setup_cube):
    ngroups = 7
    nframes = 1
    data, gdq, nframes, read_noise, rej_threshold = setup_cube(ngroups, readnoise=5)
    # if the first and last group are used in jump detection it would cause
    # jumps to be detected between group 0-1 and group 5-6
    data[0, 0, 100, 100] = 10000.0
    data[0, 1, 100, 100] = 30000.0
    data[0, 2, 100, 100] = 30020.0
    data[0, 3, 100, 100] = 40040.0  # jump
    data[0, 4, 100, 100] = 40060.0
    data[0, 5, 100, 100] = 40080.0
    data[0, 6, 100, 100] = 50000.0
    # set the first and last group to DO_NOT_USE
    gdq[0, 0, 100, 100] = dqflags.group['DO_NOT_USE']
    gdq[0, 6, 100, 100] = dqflags.group['DO_NOT_USE']
    find_crs(data, gdq, read_noise, rej_threshold, nframes)
    assert gdq[0, 0, 100, 100] == dqflags.group['DO_NOT_USE']
    assert gdq[0, 6, 100, 100] == dqflags.group['DO_NOT_USE']
    assert gdq[0, 3, 100, 100] == dqflags.group['JUMP_DET']
```

Reply:
I have a question on the setup_cube. Does it not set up a 2048 X 2048 array of pixels.
Should be make the setup_cube be smaller if you don't want to run find_crs over all
the pixels but a subset ?

Reply:
Until earlier today it did setup 2048x2048 images. Now they're only 204x204.

Reply:
I re-did the test by putting it in the existing test_twopoint_difference.py
I have the code that James has above. I added readnoise = 25 (because MIRI has a read noise ~ 25 electrons) 
I ran the test and the jump that I inserted just as a secondarytest (between groups 3 and 4) is not detected. Hum ?  I printed out the read_noise array - it is 25, the rej_threshold is 3.
Should I push this over to see if you can see anything wrong ?

Reply:
do I have the read noise in the wrong units ? I don't see the gain in this test

Reply:
changed the read_noise to 5 still no jump detection